driver-sync/src/main/com/mongodb/client/internal/KeyRetriever.java

```diff
@@ -28,19 +26,19 @@
 import static com.mongodb.assertions.Assertions.notNull;
 
 class KeyRetriever implements Closeable {
-    private final MongoClient client;
+    private final SimpleMongoClient client;
     private final boolean ownsClient;
-    private final MongoCollection<BsonDocument> collection;
+    private final MongoNamespace namespace;
 
-    KeyRetriever(final MongoClient client, final boolean ownsClient, final MongoNamespace namespace) {
-        notNull("namespace", namespace);
+    KeyRetriever(final SimpleMongoClient client, final boolean ownsClient, final MongoNamespace namespace) {
         this.client = notNull("client", client);
         this.ownsClient = ownsClient;
-        this.collection = client.getDatabase(namespace.getDatabaseName()).getCollection(namespace.getCollectionName(), BsonDocument.class);
+        this.namespace = notNull("namespace", namespace);
```

Comment:
OK, this is a hack.  The reason I save the namespace here and call getDatabase in the find method below is that at the time this constructor is called, the MongoClient has not been fully constructed yet.  In particular, the `delegate` field is not initialized, which is needed for the `getDatabase` method.  

I thought this was ok since it's all internal, but let me know if it's too ugly to bear.

Reply:
No its fine :)